### DISCUSSION TITLE
Best practice to separate tsp file

### DISCUSSION THREAD

**Discussion by muraaano** [microsoft/typespec#5242] [Q&A]
Just I started to use TypeSpec in my project. Currently tsp files are separated for each models.
But I am wondering models to be in one file or not.

Pros.
- No file handling needed(file create, delete, move)

Cons.
- One huge file possibly created

I would appreciate someone give me any advice / experience for tsp file separation. Also any file structure for TypeSpec is welcome.

**Comment by allenjzhang** [comment]
The organization of files is largely a matter of personal preference and the project size. I aim to strike a balance between the extremes of having one file per model and of putting everything into a single file. 

Here's my general approach:

- Use main.tsp for the main namespace declaration and service definition. If the service is **very small**, I prefer to keep everything in main.tsp.

- Group interfaces and related resources together.
For multiple interfaces, I place each interface and its related resource (not sub-models) in its own file. Sub-types, enums, unions, and other related types are organized in a separate models.tsp file. This setup keeps the REST resource and its route together for easier readability.
```
main.tsp
pet.tsp   (contains only top model Pet and interface Pets)
owner.tsp
models.tsp  (all other referenced sub models, enums, unions etc)
```
- Organize sub-services into separate folders for larger projects.
If the project includes multiple sub-services that are significantly different but still belong to a single TypeSpec project, I group them into separate sub-folders.
```
main.tsp
common/
    shared.tsp
service1/
    resource1.tsp
    resource2.tsp
    models.tsp
service2/
    resource1.tsp
    resource2.tsp
    models.tsp
```

Ultimately, there’s no right or wrong way to organize files—choose what works best for your project. I hope this approach provides some useful ideas!
